crates/xline/src/conflict/tests.rs

```diff
@@ -0,0 +1,221 @@
+    LeaseRevokeRequest, PutRequest, RequestWrapper,
+};
+
+use crate::conflict::{
+    spec_pool::ExclusiveSpecPool,
+    uncommitted_pool::{ExclusiveUncomPool, KvUncomPool, LeaseUncomPool},
+};
+
+use super::spec_pool::{KvSpecPool, LeaseSpecPool};
+
+#[test]
+fn kv_sp_operations_is_ok() {
+    let mut sp = KvSpecPool::default();
+    let mut gen = EntryGenerator::default();
+    let entry1 = gen.gen_put("a");
+    let entry2 = gen.gen_put("b");
+    let entry3 = gen.gen_delete_range("a", "d");
+    let entry4 = gen.gen_delete_range("c", "e");
+    assert!(sp.insert_if_not_conflict(entry1.clone()).is_none());
+    assert!(sp.insert_if_not_conflict(entry1.clone()).is_some());
+    assert!(sp.insert_if_not_conflict(entry2.clone()).is_none());
+    assert!(sp.insert_if_not_conflict(entry3.clone()).is_some());
+    assert!(sp.insert_if_not_conflict(entry4.clone()).is_none());
+    assert_eq!(sp.all().len(), 3);
+    assert_eq!(sp.len(), 3);
+    sp.remove(entry1.clone());
+    assert!(sp.insert_if_not_conflict(entry3.clone()).is_some());
+    sp.remove(entry2.clone());
+    assert!(sp.insert_if_not_conflict(entry3.clone()).is_some());
+    sp.remove(entry4.clone());
+    assert!(sp.insert_if_not_conflict(entry3.clone()).is_none());
+    sp.clear();
+    assert!(sp.is_empty());
+    assert_eq!(sp.len(), 0);
+}
+
+#[test]
+fn kv_ucp_operations_is_ok() {
+    let mut ucp = KvUncomPool::default();
+    let mut gen = EntryGenerator::default();
+    let entry1 = gen.gen_put("a");
+    let entry2 = gen.gen_put("a");
+    let entry3 = gen.gen_put("b");
+    let entry4 = gen.gen_delete_range("a", "d");
+    let entry5 = gen.gen_delete_range("c", "e");
+    let entry6 = gen.gen_delete_range("c", "f");
+    assert!(!ucp.insert(entry1.clone()));
+    assert!(ucp.insert(entry2.clone()));
+    assert!(!ucp.insert(entry3.clone()));
+    assert!(ucp.insert(entry4.clone()));
+    assert!(ucp.insert(entry5.clone()));
+    assert_eq!(ucp.all().len(), 5);
+    assert_eq!(ucp.len(), 5);
+    assert_eq!(ucp.all_conflict(&entry1).len(), 3);
+    assert_eq!(ucp.all_conflict(&entry6).len(), 2);
+    ucp.remove(entry4.clone());
+    ucp.remove(entry5.clone());
+    assert!(!ucp.insert(entry6.clone()));
+    ucp.clear();
+    assert!(ucp.is_empty());
+    assert_eq!(ucp.len(), 0);
+}
+
+#[test]
+fn lease_sp_operations_is_ok() {
+    let mut sp = LeaseSpecPool::default();
+    let mut gen = EntryGenerator::default();
+    let entry1 = gen.gen_lease_grant(0);
+    let entry2 = gen.gen_lease_grant(1);
+    let entry3 = gen.gen_lease_revoke(0);
+    let entry4 = gen.gen_lease_revoke(1);
+    assert!(sp.insert_if_not_conflict(entry1.clone()).is_none());
+    assert!(sp.insert_if_not_conflict(entry1.clone()).is_some());
+    assert!(sp.insert_if_not_conflict(entry2.clone()).is_none());
+    assert!(sp.insert_if_not_conflict(entry3.clone()).is_some());
+    assert!(sp.insert_if_not_conflict(entry4.clone()).is_some());
+    assert_eq!(sp.all().len(), 2);
```

Comment:
What's the difference between the `sp.len()` and the `sp.all().len()`

Reply:
They have the same meaning